### PR TITLE
[Launcher]: Refine GPU/Port related codes

### PR DESCRIPTION
**Pull request by yqwang-ms**
1. [Launcher]: Fail Fast if Hadoop does support GPU/Port but user specified
2. [Launcher]: Add more validations for GPU/Port
3. [Launcher]: Fix bugs for GPU/Port related codes
4. [Launcher]: Do not expose internal fields outside

### CODE REVIEW

File: frameworklauncher/src/test/resources/TestInputs/WrongFrameworkDescription.json [hunk 1]
```diff
@@ -48,39 +37,45 @@
...
         ],
-        "diskType": 0,
-        "diskMB": 31
+        "resource": {
+          "cpuNumber": 28,
+          "memoryMB": 29,
+          "gpuNumber": 2,
+          "gpuAttribute": 1,
+          "portDefinitions": {
+            "ht%%tp": null,
```
Comment (qinchen123) [inline frameworklauncher/src/test/resources/TestInputs/WrongFrameworkDescription.json]: >%% [](start = 15, length = 2)

what ?

Comment (yqwang-ms) [reply frameworklauncher/src/test/resources/TestInputs/WrongFrameworkDescription.json]: This is wrong input used to validate test

---
In reply to: [189181219](https://github.com/Microsoft/pai/pull/588#discussion_r189181219) [](ancestors = 189181219)

File: frameworklauncher/doc/USERMANUAL.md [hunk 2]
```diff
@@ -587,7 +587,7 @@ Launcher sets up below EnvironmentVariables for each User Service to use:
 | EnvironmentVariable | Description |
 |:---- |:---- |
 | CONTAINER_GPUS | Only valid when gpuNumber is greater than 0. It is a number, each bit of this number represents a Gpu, for example, 3 represents gpu0 and gpu1. |
-| CONTAINER_PORTS | Only valid when portDefinitions is not empty. It is a string, a format example is: "portLabel1:port1,port2,port3;portLabel2:port4,port5,port6". |
+| CONTAINER_PORTS | Only valid when portDefinitions is not empty. It is a string, a format example is: "portLabel1:port1,port2;portLabel2:port4,port5;". |
```
Comment (qinchen123) [inline frameworklauncher/doc/USERMANUAL.md]: >; [](start = 126, length = 1)

where is port3 ?  1,2, 4, 5 ?

Comment (yqwang-ms) [reply frameworklauncher/doc/USERMANUAL.md]: OK

---
In reply to: [189219910](https://github.com/Microsoft/pai/pull/588#discussion_r189219910) [](ancestors = 189219910)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java [hunk 3]
```diff
@@ -255,18 +257,18 @@ public synchronized SelectionResult select(ResourceDescriptor requestResource, S
...
-        optimizedRequestResource.setPortRanges(portRanges);
+      if (ValueRangeUtils.getValueNumber(reusedPorts) > 0) {
+        LOGGER.logInfo(
+            "select: reuse pre-selected ports: %s", CommonExts.toString(reusedPorts));
+        optimizedRequestResource.setPortRanges(reusedPorts);
+      } else {
+        List<ValueRange> portRanges = selectPortsFromFilteredNodes(optimizedRequestResource);
+        LOGGER.logInfo(
+            "select: select ports from all filteredNodes: %s", CommonExts.toString(portRanges));
+        if (ValueRangeUtils.getValueNumber(portRanges) >= optimizedRequestResource.getPortNumber()) {
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java]: port in PortRanges must equal with getPortNubmer(), If it is >, there are some issue in selectPortsFromFilteredNodes()

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java]: OK

---
In reply to: [189478035](https://github.com/Microsoft/pai/pull/588#discussion_r189478035) [](ancestors = 189478035)

**Pull request merged**